README.md
# modbusd

[![Build Status](http://dev.cmwang.net/api/badges/taka-wang/modbusd/status.svg)](http://dev.cmwang.net/taka-wang/modbusd)

Modbus master daemon

- Support doxygen style comments.
- ZeroMQ is a high-level message library, you can replace it with your own data bus implementations without losing the core functionalities.

## Table of content

- [Design](#design)
- [Setup](#setup)
- [Continuous Integration](#ci)
- [Documentation](#doc)

---

<a name="design"></a>

## Design

### Implemented libmodbus function codes

>| FC    | Description            | #Len    | API                                                                                                                                                 |
>|:-----:|------------------------|---------|-----------------------------------------------------------------------------------------------------------------------------------------------------|
>| 0x01  | read coils             |  2000   |[int modbus_read_bits(modbus_t *ctx, int addr, int nb, uint8_t *dest)](http://libmodbus.org/docs/v3.1.4/modbus_read_bits.html)                       |
>| 0x02  | read discrete inputs   |  2000   |[int modbus_read_input_bits(modbus_t *ctx, int addr, int nb, uint8_t *dest)](http://libmodbus.org/docs/v3.1.4/modbus_read_input_bits.html)           |
>| 0x03  | read holding registers |  125    |[int modbus_read_registers(modbus_t *ctx, int addr, int nb, uint16_t *dest)](http://libmodbus.org/docs/v3.1.4/modbus_read_registers.html)            |
>| 0x04  | read input registers   |  125    |[int modbus_read_input_registers(modbus_t *ctx, int addr, int nb, uint16_t *dest)](http://libmodbus.org/docs/v3.1.4/modbus_read_input_registers.html)|
>| 0x05  | write single coil      |   -     |[int modbus_write_bit(modbus_t *ctx, int addr, int status)](http://libmodbus.org/docs/v3.1.4/modbus_write_bit.html)                                  |
>| 0x06  | write single register  |   -     |[int modbus_write_register(modbus_t *ctx, int addr, int value)](http://libmodbus.org/docs/v3.1.4/modbus_write_register.html)                         |
>| 0x0F  | write multi coils      |  1968   |[int modbus_write_bits(modbus_t *ctx, int addr, int nb, const uint8_t *src)](http://libmodbus.org/docs/v3.1.4/modbus_write_bits.html)                |
>| 0x10  | write multi registers  |  125    |[int modbus_write_registers(modbus_t *ctx, int addr, int nb, const uint16_t *src)](http://libmodbus.org/docs/v3.1.4/modbus_write_registers.html)     |

### Coil/register number and address table

>|Coil/Register numbers|data address       |type          |table name                     |offset| function code|
>|:--------------------|:------------------|:-------------|:------------------------------|:-----|:-------------|
>|1-9999               |0000 to 270E (9998)|Read-Write    |Discrete Output Coils          |1     | 1, 5, 15     |
>|10001-19999          |0000 to 270E (9998)|Read-Only     |Discrete Input Contacts        |10001 | 2            |
>|30001-39999          |0000 to 270E (9998)|Read-Only     |Analog Input Registers         |30001 | 4            |
>|40001-49999          |0000 to 270E (9998)|Read-Write    |Analog Output Holding Registers|40001 | 3, 6, 16     |

### Command mapping table

>| Command         | Number | Description      |
>|:---------------:|-------:|:-----------------|
>| fc1             |   1    | modbus fc 1      |
>| fc2             |   2    | modbus fc 2      |
>| fc3             |   3    | modbus fc 3      |
>| fc4             |   4    | modbus fc 4      |
>| fc5             |   5    | modbus fc 5      |
>| fc6             |   6    | modbus fc 6      |
>| fc15            |  15    | modbus fc 15     |
>| fc16            |  16    | modbus fc 16     |
>| set_tcp_timeout |  50    | set tcp timeout  |
>| get_tcp_timeout |  51    | get tcp timeout  |


---

### Configuration File

```javascript
{
    "syslog": 1,
    "zmq":
    {
        "sub": "ipc:///tmp/to.modbus",
        "pub": "ipc:///tmp/from.modbus"
    },
    "mbtcp":
    {
        "connect_timeout": 200000
    }
}
```

### Modbus TCP command format

Please refer to [command definition](command.md).

### External libraries

- [libmodbus](http://libmodbus.org)
- [libzmq](https://github.com/zeromq/libzmq)
- [czmq](https://github.com/zeromq/czmq)
- [uthash](https://troydhanson.github.io/uthash)
- [cJSON](https://github.com/DaveGamble/cJSON)

---

### Library documentations

- [uthash user guide](http://troydhanson.github.io/uthash/userguide.html)
- [libmodbus api document](http://libmodbus.org/docs/v3.1.4/)
- [libmodbus header](https://github.com/stephane/libmodbus/blob/master/src/modbus.h)
- [cJSON examples](https://github.com/DaveGamble/cJSON)


### Flow Chart

![flow](image/flow.png)

---

<a name="setup"></a>

## Setup

### Setup development dependencies

```bash
sudo apt-get update
sudo apt-get install -y git build-essential autoconf libtool pkg-config cmake
```

---

### Setup OSS libs dependencies

#### Install libmodbus library (3.1.4)

```bash
git clone https://github.com/stephane/libmodbus/
cd libmodbus
./autogen.sh
./configure
make
sudo make install
sudo ldconfig
```

#### Install libzmq (3.2.5)

```bash
wget https://github.com/zeromq/zeromq3-x/releases/download/v3.2.5/zeromq-3.2.5.tar.gz
tar xvzf zeromq-3.2.5.tar.gz
cd zeromq-3.2.5
./configure
make
sudo make install
sudo ldconfig
```

#### Install czmq (high-level C binding for zeromq)

```bash
git clone git://github.com/zeromq/czmq.git
cd czmq
./autogen.sh
./configure
make
sudo make install
sudo ldconfig
```

### Build from source code

```bash
git clone modbusd
cd modbusd
mkdir build
cd build
cmake ..
make
./modbusd ../modbusd.json # load external configuration file
```

---

<a name="ci"></a>

## Continuous Integration

I do [continuous integration](test) and update docker images after git push by self-hosted drone.io server and dockerhub service [~~Travis CI~~](https://travis-ci.org/taka-wang/modbusd).

## Test Cases

- [x] Test holding registers (4x)
    - [x] `4X Table: 60000` Read/Write uint16 value test: FC6, FC3
    - [x] `4X Table: 30000` Read/Write int16 value test: FC6, FC3
    - [x] `4X Table` Multiple read/write test: FC16, FC3
    - [x] `4X Table` Multiple read/write test: FC16, FC3
- [x] Test coils (0x)
    - [x] `0X Table` Single read/write test:FC5, FC1
    - [x] `0X Table` Multiple read/write test: FC15, FC1
- [x] Test Discrete Input (1x)
    - [x] `1X Table` read test: FC2
- [x] Test Input Registers (3x)
    - [x] `3X Table` read test:FC4
- [x] Test TCP Timeout
    - [x] `Set timeout` test
    - [x] `Get timeout` test

![ci](image/ci-drone.png)

### Images and testing from the scratch

```bash

docker pull takawang/c-modbus-slave:x86
docker build -t takawang/modbusd .
docker build -t takawang/dummy-psmbtcp test/dummy-psmbtcp/.

docker run -itd --name=slave takawang/c-modbus-slave:x86
docker run -v /tmp:/tmp --link slave -it --name=modbusd takawang/modbusd
docker run -v /tmp:/tmp -it --link slave takawang/dummy-psmbtcp
```

### Docker compose

```bash
docker-compose up --abort-on-container-exit
```

### Deployment Diagram

![Deployment](http://plantuml.com/plantuml/svg/5Sl13O0W343HLNG0wTrjAcs0I2c1DiRjnVD_VoyjLYVsKRTirkS9CF09gLZsooUFgCsuMOWgO7ZZyM1Bq5qg24xZ0SIzwYiBWIYjYSAVFm00)

---

<a name="doc"></a>

## Documentations

- [API Documentation](http://taka-wang.github.io/modbusd)


HISTORY.md
# History

## v0.5.0

- [x] Refine test cases

## v0.4.3

### Done

- [x] Replace modbusd base image with alpine (for production)
- [x] Replace modbus slave test server base image with alpine

## v0.4.0

### Done

- [x] rewrite goclient according to psmb's types
- [x] refactor response zframe definition: [cmd, json]

## v0.3.0

### Done

- [x] rewrite test server in c, test cases in golang
- [x] update travis ci with new test env
- [x] implement timeout.set, timeout.get functions
- [x] implement json_get_double, json_set_double, json_get_long
- [x] refactor tid data type
- [x] refactor command definition

---

## v0.2.1

### Done

- [x] fix uthash sizeof issue (1000 items); cause: memset to the wrong size
- [x] fix char pointer key issue (hash); use char array instead of char pointer (unkown length)
- [x] modbus_connect hang issue; set tcp timeout
- [x] handle 'reset by peer' issue; workaround: set connection flag to false :warning:
- [x] implement keep connection mechanism via hashtable
- [x] implement FC (1~6, 15, 16)
- [x] assign daemon version number from the latest git tag
- [x] implement syslog and flag mechanism
- [x] implement read/write config mechanism
- [x] define default config
- [x] implement dummy modbus server in node.js for testing
- [x] support ipv4/v6 ip address string
- [x] refactor int port to char * port
- [x] support docker compose
- [x] support valgrind (disable now)
- [x] implement set/get timeout command
- [x] support mocha and async test (runs slow on cloud server)
- [x] add versioneye support (depends check)
- [x] support armhf
- [x] deploy doxygen document automatically

---

## Backlog

- [ ] implement [circuit breaker pattern](http://martinfowler.com/bliki/CircuitBreaker.html)
- [ ] implement retry count mechanism for each modbus slave
- [ ] implement external debug flag


mbtcp.c
/**
 * @file mbtcp.c
 * @author Taka Wang
 * @brief modbus tcp functions
*/

#include "mb.h"

/* ==================================================
 *  global variable
================================================== */

extern int enable_syslog;                   // syslog flag
static mbtcp_handle_s *mbtcp_htable = NULL; // hashtable header
long tcp_conn_timeout_usec = 200000;        /* tcp connection
                                               timeout in usec */

/* ==================================================
 *  static functions
================================================== */

/**
 * @brief Combo function: get or init mbtcp handle.
 *
 * @param ptr_handle Pointer to mbtcp handle.
 * @param req cJSON request object.
 * @return Success or not.
 */
static bool lazy_init_mbtcp_handle(mbtcp_handle_s **ptr_handle, cJSON *req)
{
    BEGIN(enable_syslog);
    
    char *ip   = json_get_char (req, "ip");
    char *port = json_get_char (req, "port");
    
    if (mbtcp_get_handle (ptr_handle, ip, port))
    {
       return true;
    }
    else
    {
        if (mbtcp_init_handle(ptr_handle, ip, port))
        {
            return true;
        }
        else
        {
            // Fatal! Unable to allocate mbtcp context,
            // maybe system resourse issue!
            return false;
        }
    }
}

/**
 * @brief Combo function: check mbtcp connection status, 
 *        if not connected, try to connect to slave.
 *
 * @param handle Mbtcp handle.
 * @param reason Pointer to fail reason string.
 * @return Success or not.
 */
static bool lazy_mbtcp_connect(mbtcp_handle_s *handle, char **reason)
{
    BEGIN(enable_syslog);
        
    if (mbtcp_get_connection_status(handle))
    {
        return true;
    }
    else
    {
        if (mbtcp_do_connect(handle, reason))
        {
             return true;
        }
        else
        {
            // could get fail reason from '*reason'
            return false;
        }
    }   
}

/**
 * @brief Help function. FC1, FC2 request handler
 *
 * @fc Function code 1 and 2 only.
 * @param handle Mbtcp handle.
 * @param req cJSON request object.
 * @return Modbus response string in JSON format.
 */
static char * mbtcp_read_bit_req(uint8_t fc, mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    int addr = json_get_int(req, "addr");
    int len  = json_get_int(req, "len");
    char * tid = json_get_char(req, "tid");
    if (len > MODBUS_MAX_READ_BITS) // 2000
    {
        return set_modbus_fail_resp_str(tid, "Too many bits requested");
    }
    else
    {
        uint8_t bits[len];
        // memory reset for variable length array
        memset(bits, 0, len * sizeof(uint8_t));
        int ret = 0;
        switch (fc)
        {
            case 1:
                ret = modbus_read_bits(handle->ctx, addr, len, bits);
                break;
            case 2:
                ret = modbus_read_input_bits(handle->ctx, addr, len, bits);
                break;
            default:
                return set_modbus_fail_resp_str(tid, "Wrong function code");
        }

        if (ret < 0) 
        {
            return set_modbus_fail_resp_str_with_errno(tid, handle, errno);
        } 
        else 
        {
            LOG(enable_syslog, "fc:%d, desired length: %d, read length:%d", fc, len, ret);
            
            // [TODO]:remove; debug only
            for (int ii = 0; ii < ret; ii++) 
            {
                LOG(enable_syslog, "[%d]=%d", ii, bits[ii]);
            }

            // uint8_t array
            return set_modbus_success_resp_str_with_data(tid, cJSON_CreateUInt8Array(bits, len));
        }
    }    
}

/**
 * @brief Help function. FC3, FC4 request handler
 *
 * @fc Function code 3 and 4 only.
 * @param handle Mbtcp handle.
 * @param req cJSON request object.
 * @return Modbus response string in JSON format.
 */
static char * mbtcp_read_reg_req(uint8_t fc, mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    int addr = json_get_int(req, "addr");
    int len  = json_get_int(req, "len");
    char * tid = json_get_char(req, "tid");
    if (len > MODBUS_MAX_READ_REGISTERS) // 125
    {
        return set_modbus_fail_resp_str(tid, "Too many registers requested");
    }
    else
    {
        uint16_t regs[len];
        // memory reset for variable length array
        memset(regs, 0, len * sizeof(uint16_t));
        int ret = 0;
        switch (fc)
        {
            case 3:
                ret = modbus_read_registers(handle->ctx, addr, len, regs);
                break;
            case 4:
                ret = modbus_read_input_registers(handle->ctx, addr, len, regs);
                break;
            default:
                return set_modbus_fail_resp_str(tid, "Wrong function code");
        }
        
        if (ret < 0) 
        {
            return set_modbus_fail_resp_str_with_errno(tid, handle, errno);
        } 
        else 
        {
            LOG(enable_syslog, "fc:%d, desired length: %d, read length:%d", fc, len, ret);
            
            // [TODO]:remove; debug only
            for (int ii = 0; ii < ret; ii++) 
            {
                LOG(enable_syslog, "[%d]=%d", ii, regs[ii]);
            }
            
            // uint16_t array
            return set_modbus_success_resp_str_with_data(tid, cJSON_CreateUInt16Array(regs, len));
        }
    }
}

/**
 * @brief Help function. FC5, FC6 request handler
 *
 * @fc Function code 5 and 6 only.
 * @param handle Mbtcp handle.
 * @param req cJSON request object.
 * @return Modbus response string in JSON format.
 */
static char * mbtcp_single_write_req(uint8_t fc, mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    int addr = json_get_int(req, "addr");
    char * tid = json_get_char(req, "tid");
    int data = json_get_int(req, "data");
    int ret  = 0;
    switch (fc)
    {
        case 5:
            ret = modbus_write_bit(handle->ctx, addr, data);
            break;
        case 6:
            ret = modbus_write_register(handle->ctx, addr, data);
            break;
        default:
            return set_modbus_fail_resp_str(tid, "Wrong function code");
    }

    if (ret < 0) 
    {
        return set_modbus_fail_resp_str_with_errno(tid, handle, errno);
    }
    else
    {
        return set_modbus_success_resp_str(tid);   
    }
}

/**
 * @brief Help function. FC15, FC16 request handler
 *
 * @fc Function code 15 and 16 only.
 * @param handle Mbtcp handle.
 * @param req cJSON request object.
 * @return Modbus response string in JSON format.
 */
static char * mbtcp_multi_write_req(uint8_t fc, mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    int addr = json_get_int(req, "addr");
    int len  = json_get_int(req, "len");
    char * tid = json_get_char(req, "tid");
    
    uint8_t *bits;      // FC15
    uint16_t *regs;     // FC16
    cJSON * data = NULL;
    int ret = 0;

    switch (fc)
    {
        case 15:
            // memory reset for variable length array
            bits = (uint8_t *) malloc(len * sizeof(uint8_t));
            memset(bits, 0, len * sizeof(uint8_t));
            // handle uint8_t array
            data = cJSON_GetObjectItem(req, "data");
            for (int i = 0 ; i < cJSON_GetArraySize(data) ; i++)
            {
                uint8_t subitem = cJSON_GetArrayItem(data, i)->valueint;
                bits[i] = subitem;
                LOG(enable_syslog, "[%d]=%d", i, bits[i]);
            }
            ret = modbus_write_bits(handle->ctx, addr, len, bits);
            free(bits);
            break;
        case 16:
            // memory reset for variable length array
            regs = (uint16_t *) malloc(len * sizeof(uint16_t));
            memset(regs, 0, len * sizeof(uint16_t));
            // handle uint16_t array
            data = cJSON_GetObjectItem(req, "data");
            for (int i = 0 ; i < cJSON_GetArraySize(data) ; i++)
            {
                uint16_t subitem = cJSON_GetArrayItem(data, i)->valueint;
                regs[i] = subitem;
                LOG(enable_syslog, "[%d]=%d", i, regs[i]);
            }
            ret = modbus_write_registers(handle->ctx, addr, len, regs);
            free(regs);
            break;
        default:
            return set_modbus_fail_resp_str(tid, "Wrong function code");
    }

    if (ret < 0) 
    {
        return set_modbus_fail_resp_str_with_errno(tid, handle, errno);
    } 
    else
    {
        return set_modbus_success_resp_str(tid);         
    }
}

/* ==================================================
 *  public functions
================================================== */

bool mbtcp_init_handle(mbtcp_handle_s **ptr_handle, char *ip, char *port)
{
    BEGIN(enable_syslog);

    // create a mbtcp context
    modbus_t *ctx = modbus_new_tcp_pi(ip, port);
    
    if (ctx == NULL)
    {
        ERR(enable_syslog, "Unable to allocate mbtcp context");
        return false; // fail to allocate context
    }
        
    // set tcp connection timeout
    modbus_set_response_timeout(ctx, 0, tcp_conn_timeout_usec);
    LOG(enable_syslog, "set response timeout:%ld", tcp_conn_timeout_usec);
    
    // @add context to mbtcp hashtable
    mbtcp_handle_s *handle;
    handle = (mbtcp_handle_s*)malloc(sizeof(mbtcp_handle_s));
    // let alignment bytes being set to zero-value!!
    memset(handle, 0, sizeof(mbtcp_handle_s));
    handle->connected = false;
    strcpy(handle->key.ip, ip);
    strcpy(handle->key.port, port);
    handle->ctx = ctx;

    HASH_ADD(hh, mbtcp_htable, key, sizeof(mbtcp_key_s), handle);
    LOG(enable_syslog, "Add %s:%s to mbtcp hashtable", handle->key.ip, mbtcp_htable->key.port);

    // call by reference to `mbtcp handle address`
    *ptr_handle = handle;

    // @connect to server
    char * reason = NULL;
    mbtcp_do_connect(handle, &reason);
    return true;
}

bool mbtcp_get_handle(mbtcp_handle_s **ptr_handle, char *ip, char *port)
{
    BEGIN(enable_syslog);
    
    mbtcp_handle_s query, *hash_ctx;
    memset(&query, 0, sizeof(mbtcp_handle_s));
    strcpy(query.key.ip, ip);
    strcpy(query.key.port, port);
    HASH_FIND(hh, mbtcp_htable, &query.key, sizeof(mbtcp_key_s), hash_ctx);
    
    if (hash_ctx != NULL)
    {
        LOG(enable_syslog, "tcp server %s:%s found", hash_ctx->key.ip, hash_ctx->key.port);
        // call by reference to `mbtcp handle address`
        *ptr_handle = hash_ctx; 
        return true;
    }
    else
    {
        ERR(enable_syslog, "tcp server %s:%s not found", query.key.ip, query.key.port);
        *ptr_handle = NULL; 
        return false; // not found
    }
}

void mbtcp_list_handles() 
{
    BEGIN(enable_syslog);
    mbtcp_handle_s * handle;

    for (handle = mbtcp_htable; handle != NULL; handle = handle->hh.next)
    {
        printf("ip:%s, port:%s\n", handle->key.ip, handle->key.port);
    }
    END(enable_syslog);
}

bool mbtcp_do_connect(mbtcp_handle_s *handle, char ** reason)
{
    BEGIN(enable_syslog);
    
    if (handle == NULL)
    {
        ERR(enable_syslog, "NULL handle");
        *reason = "NULL handle";
        return false;
    }
    
    if (modbus_connect(handle->ctx) == -1) 
    {
        ERR(enable_syslog, "Connection failed: %s", modbus_strerror(errno));
        *reason = (char *)modbus_strerror(errno);
        return false;
    }
    else
    {
        LOG(enable_syslog, "%s:%s connected", handle->key.ip, handle->key.port);
        handle->connected = true;
        return true;
    }
}

bool mbtcp_get_connection_status(mbtcp_handle_s *handle)
{
    BEGIN(enable_syslog);
    
    if (handle == NULL)
    {
        ERR(enable_syslog, "NULL handle");
        return false;
    }
    
    LOG(enable_syslog, "%s:%s connected: %s", handle->key.ip, 
                                              handle->key.port, 
                                              handle->connected ? "true" : "false");
    return handle->connected;
}

char * mbtcp_cmd_hanlder(cJSON *req, mbtcp_fc fc)
{
    BEGIN(enable_syslog);
    mbtcp_handle_s *handle = NULL;
    char * tid = json_get_char(req, "tid");
    if (lazy_init_mbtcp_handle(&handle, req))
    {
        char * reason = NULL;
        if (lazy_mbtcp_connect(handle, &reason))
        {
            // set slave id
            int slave = json_get_int(req, "slave");
            LOG(enable_syslog, "slave id: %d", slave);
            modbus_set_slave(handle->ctx, slave);
            return fc(handle, req);
        }
        else
        {
            // [enhance]: get reason from modbus response
            return set_modbus_fail_resp_str(tid, reason);
        }
    }
    else
    {
        return set_modbus_fail_resp_str(tid, "init modbus tcp handle fail");
    }
}

char * mbtcp_set_response_timeout(char * tid, long timeout)
{
    BEGIN(enable_syslog);

    // set timeout
    tcp_conn_timeout_usec = timeout;
    
    cJSON *resp_root;
    resp_root = cJSON_CreateObject();
    cJSON_AddStringToObject(resp_root, "tid", tid);
    cJSON_AddStringToObject(resp_root, "status", "ok");
    char * resp_json_string = cJSON_PrintUnformatted(resp_root);
    LOG(enable_syslog, "resp: %s", resp_json_string);
    
    // clean up
    cJSON_Delete(resp_root);
    END(enable_syslog);
    return resp_json_string;
}

char * mbtcp_get_response_timeout(char * tid)
{
    BEGIN(enable_syslog);

    cJSON *resp_root;
    resp_root = cJSON_CreateObject();
    cJSON_AddStringToObject(resp_root, "tid", tid);
    cJSON_AddNumberToObject(resp_root, "timeout", tcp_conn_timeout_usec);
    cJSON_AddStringToObject(resp_root, "status", "ok");
    char * resp_json_string = cJSON_PrintUnformatted(resp_root);
    LOG(enable_syslog, "resp: %s", resp_json_string);
    
    // clean up
    cJSON_Delete(resp_root);
    END(enable_syslog);
    return resp_json_string;
}

// Read coils.
char * mbtcp_fc1_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_read_bit_req(1, handle, req);
}

// Read discrete input.
char * mbtcp_fc2_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_read_bit_req(2, handle, req);
}

// Read holding registers.
char * mbtcp_fc3_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);   
    return mbtcp_read_reg_req(3, handle, req);  
}

// Read input registers.
char * mbtcp_fc4_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_read_reg_req(4, handle, req);  
}

// Write single coil.
char * mbtcp_fc5_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_single_write_req(5, handle, req);
}

// Write single register.
char * mbtcp_fc6_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_single_write_req(6, handle, req);
}

// Write multiple coils.
char * mbtcp_fc15_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_multi_write_req(15, handle, req);
}

// Write multiple registers.
char * mbtcp_fc16_req(mbtcp_handle_s *handle, cJSON *req)
{
    BEGIN(enable_syslog);
    return mbtcp_multi_write_req(16, handle, req);
}